python: add config file option

loicreynier opened the pull request:
Not sure if you are open to contributions but if you are I added an option to the Python module for the configuration of the [interpreter startup file](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONSTARTUP).

loicreynier committed:
python: add config file option

schuelermine commented:
I will test this this evening, thank you!

schuelermine commented:
If you don’t mind I or you could add a few other useful options? Specifically 3.13 will add the [`PYTHON_HISTORY`](https://docs.python.org/3.13/using/cmdline.html#envvar-PYTHON_HISTORY) and [`PYTHON_COLORS`](https://docs.python.org/3.13/using/cmdline.html#envvar-PYTHON_COLORS) environment variables setting which globally might be desired.
However that might be a premature addition, feel free to chime in.

loicreynier commented:
I actually did not these environment variables existed. I used to implement a custom Python history location with my own startup functions. I'll try them out during the week and I'll add them to the module.

schuelermine commented:
Well they don’t exist yet, they’ll be added in 3.13

loicreynier committed:
python: add options for Python 3.13 vars

loicreynier commented:
I added an option for both variables.

The environment variables seem to have no effect (yet) on the Python 3.13 build available in `nixpkgs-unstable`, so I could not properly test them.

I could have added a test raising an error if the options are used with a Python version older than 3.13, but since these variables do not conflict with previous versions, I thought that would be overkill.

schuelermine commented:
OK, I will test this soon-ish and then merge. I am currently very busy with Uni so I unfortunately don’t have much time!

schuelermine commented on languages/python/default.nix:
```diff
@@ -49,6 +49,79 @@ in {
+        for details.
+      '';
+      default = "${config.xdg.configHome}/python/startup.py";
+      defaultText = "$XDG_CONFIG_HOME/python/startup.py";
```
It’s weird to use `$` interpolation in the example text when you actually use Nix antiquotation based on the option value, since `config.sessionVariables` _would_ interpolate that automatically, but only if `$XDG_CONFIG_HOME` is defined.

schuelermine replied:
Maybe leave this `null` by default.

schuelermine commented on languages/python/default.nix:
```diff
@@ -49,6 +49,79 @@ in {
+      '';
+    };
+    enableColors =
+      lib.mkEnableOption ''
+        colors in the interpreter. See
```
It seems the text here is not complete.

schuelermine replied:
I’m not sure if this should even be a `mkEnableOption`. You _can_ leave the variable undefined which I presume makes it detect a terminal to enable color (the docs page is unclear on this—maybe try it out once there’s a build with colors).

schuelermine replied:
The text should probably include a mention of the alternative global variables `NO_COLOR` and `FORCE_COLOR`, which the user can define themselves.

schuelermine commented on languages/python/default.nix:
```diff
@@ -49,6 +49,79 @@ in {
       defaultText = lib.literalExpression "pkgs: [ ]";
       example = lib.literalExpression "pkgs: [ pkgs.requests ]";
     };
+    config = lib.mkOption {
```
Are you sure we want to call it “config”? Granted I also have `ghciConfig` but maybe that should be renamed too, now that I think about it.

loicreynier replied:
Maybe we could name it `initExtra` (or `init` since the module doesn't write anything to the file) to match the other shell HM options (bash, zsh...)?

loicreynier replied:
By the way, it seems that in HM the convention is to name `configFile` the variable for the configuration file path. We could rename `configPath` to `configFile` too.

schuelermine replied:
Good idea.

schuelermine commented on languages/python/default.nix:
```diff
@@ -49,6 +49,79 @@ in {
+    ];
+    file."${cfg.configPath}" = lib.mkIf (cfg.config != null) {
+      text = ''
+        """Python interpreter configuration file.
+
+        Generated by Home Manager, do not edit manually.
+        See: https://docs.python.org/3/using/cmdline.html#envvar-PYTHONSTARTUP
+
+        This script is executed at the start of each interactive python console
+        when it is defined in the `PYTHONSTARTUP` environment variable.
+        """
```
I suggest the following wording:
```suggestion
        # DO NOT EDIT -- this file has been generated automatically.
        # Python interpreter startup commands, generated via home-manager.
```

In particular I would avoid using a multiline string as a comment here as it gets picked up by Python as a docstring and gets stored as the `__doc__` variable.

schuelermine commented on languages/python/default.nix:
```diff
@@ -49,6 +49,79 @@ in {
+      ''
+      // {default = true;};
+  };
+  config.home = {
+    packages = lib.mkIf cfg.enable [cfg.package];
+    sessionVariables = lib.mkMerge [
+      (lib.mapAttrs
+        (_: toString)
+        (lib.filterAttrs (_: var: var != [] && var != null) {
+          PYTHONSTARTUP = cfg.configPath;
+          PYTHON_HISTORY = cfg.historyPath;
+        }))
+      (lib.mkIf (!cfg.enableColors) {PYTHON_COLORS = "-1";})
+    ];
```
Why not just something like this?
```suggestion
  config.home = {
    packages = lib.mkIf cfg.enable [cfg.package];
    sessionVariables = {
      PYTHONSTARTUP = lib.mkIf (cfg.configPath != null) (toString cfg.configPath);
      PYTHON_HISTORY = lib.mkIf (cfg.historyPath != null) (toString cfg.historyPath);
      PYTHON_COLORS = if cfg.enableColors then "0" else "1";
    };
```

schuelermine commented on languages/python/default.nix:
```diff
@@ -49,6 +49,79 @@ in {
+      ''
+      // {default = true;};
+  };
+  config.home = {
+    packages = lib.mkIf cfg.enable [cfg.package];
+    sessionVariables = lib.mkMerge [
+      (lib.mapAttrs
+        (_: toString)
+        (lib.filterAttrs (_: var: var != [] && var != null) {
+          PYTHONSTARTUP = cfg.configPath;
+          PYTHON_HISTORY = cfg.historyPath;
+        }))
+      (lib.mkIf (!cfg.enableColors) {PYTHON_COLORS = "-1";})
+    ];
```
I don’t think `"-1"` is a valid value for `PYTHON_COLORS`.

schuelermine commented on languages/python/default.nix:
```diff
@@ -49,6 +49,79 @@ in {
+      type = with lib.types; nullOr path;
+      description = ''
+        Python interpreter history file path. See
+        <https://docs.python.org/3.13/using/cmdline.html#envvar-PYTHON_HISTORY>
```
I’m not that familiar with the markdown conventions in NixOS module code but I’m also not sure if `<…>` is required/well-placed here.

loicreynier replied:
Not really familiar either but [it seems to be the convention for Home Manager](https://github.com/nix-community/home-manager/blob/b2f56952074cb46e93902ecaabfb04dd93733434/modules/programs/alacritty.nix#L45).

schuelermine replied:
Ah OK, thanks!

schuelermine commented on languages/python/default.nix:
```diff
@@ -49,6 +49,79 @@ in {
       example = lib.literalExpression "pkgs: [ pkgs.requests ]";
     };
+    config = lib.mkOption {
+      type = with lib.types; nullOr lines;
```
Maybe also give the user the option to have a path here. This could just be an alias for a `file` option.

schuelermine commented:
Also, I would prefer if you formatted your code with `nixfmt`.

schuelermine commented on languages/python/default.nix:
```diff
@@ -49,6 +49,79 @@ in {
+        }))
+      (lib.mkIf (!cfg.enableColors) {PYTHON_COLORS = "-1";})
+    ];
+    file."${cfg.configPath}" = lib.mkIf (cfg.config != null) {
```
Maybe this should be created whenever `configPath` is not `null`, or `PYTHONSTARTUP` should not be set if `config` is `null`

loicreynier replied:
In this case, wouldn't it be better to have a default value other than `null` for `configPath` by default? This would ensure the generation of a configuration file whenever `config` is not `null`. The default value could then be set via `${config.home.HomeDirectory}/.config` to avoid reliance on `xdg`.

schuelermine replied:
I agree

loicreynier committed:
python: remove config docstring

Co-authored-by: Anselm Schüler <mail@anselmschueler.com>

loicreynier committed:
python: simplify `sessionVariables`

Co-authored-by: Anselm Schüler <mail@anselmschueler.com>

loicreynier committed:
python: fix `config`

loicreynier committed:
python: format with `nixfmt`

loicreynier committed:
python: add default value for `configPath` and fix `sessionVariables`

loicreynier commented:
At this point I have not dealt with `PYTHON_COLORS` yet nor rename the options. I have re-implemented the behavior between `config` and `PYTHONSTARTUP`.
As I said above I kept a default value for `configPath` (not using `xdg`) so as the startup file and the shell variable are always set when `config` is not null.

schuelermine commented:
Hey, sorry for the late response—my local git repo got messed up.

loicreynier commented:
> Hey, sorry for the late response—my local git repo got messed up.

No worries, take your time.

schuelermine commented:
Sorry, I had a bunch of Uni work. I think this looks good, but I will test it and revise it and then merge it.

schuelermine reviewed (commented):
How about adding
```nix
{ options, config, pkgs, lib, ... }: {
```
at the top and replacing the default value for `configPath` with
```nix
"${
  if builtins.length options.xdg.configHome.definition != 0
  then config.xdg.configHome
  else "${config.home.homeDirectory}/.config"
}/python/startup.py"
```
?

schuelermine commented:
or just `if config.xdg.enable` I guess

loicreynier committed:
python: remove color option

loicreynier committed:
python: `configPath` -> `configFile`

loicreynier committed:
python: improve default value for `configFile`

loicreynier commented:
After testing with the latest build (which as color support) I decided to remove the colors options as we discussed earlier. As you said, this behavior is better managed using `NO_COLOR` since Python is able to detect if the terminal supports colors.

I renamed the `configPath` option to `configFile` as discussed earlier.
And I added your suggestion using `if config.xdg.enable` for its default value.

schuelermine committed:
python: rename, restructure, explain

schuelermine commented:
I have made some changes that I think complete this. Please give feedback. If it looks OK, I’ll merge.

On the one hand: I’m really sorry. I believe there has either been a misunderstanding or I have forgotten my old opinion and changed it, but I believe this version is better.
On the other hand: I have updated the environment variable to convert a relative path to an absolute path if needed.

loicreynier commented on languages/python/default.nix:
```diff
@@ -1,5 +1,21 @@
+    if lib.hasPrefix config.home.homeDirectory computedConfigPath then
+      computedConfigPath
+    else
+      "${config.home.homeDirectory}/${computedConfigPath}";
```
```suggestion
      (if lib.hasPrefix "~" computedConfigPath then
        "${config.home.homeDirectory}/"
        + lib.removePrefix "~" computedConfigPath
      else
        "${config.home.homeDirectory}/${computedConfigPath}");
```

Just tested, looks good to me.

I'd just add a test to remove the `~` prefix if present in `configPath`.

schuelermine replied:
I don’t like this because it means the user can’t create a file called `~` easily.

schuelermine replied:
(In this view the behavior of not duplicating a `${config.home.homeDirectory}` prefix is also questionable, but this is home-manager behavior we don’t control, and it can be easily rectified by not including the leading slash)

schuelermine commented:
Thank you very much!

schuelermine commented:
And apologies again for being so slow to respond to things.
